spharpy/spherical.py

```diff
@@ -1,6 +1,431 @@
+        the basis matrix multiplied by 4 * π * weights.
+
+        Parameters
+        ----------
+        inv_transform_type : {'pseudo_inverse', 'quadrature'}, optional
+            Type of inverse transform to compute.
+        weights : array-like, optional
+            Sampling weights for the quadrature transform. Required if 'quadrature' is chosen.
+
+        Returns
+        -------
+        None
+        """
+        if self.basis is None:
+            self.compute_basis()
+        if inv_transform_type != self.inv_transform_type:
+            assert inv_transform_type in ['pseudo_inverse', 'quadrature'], (
+                "Invalid inverse transform type, "
+                "currently only 'pseudo_inverse' "
+                "and 'quadrature' are supported")
+            self.inv_transform_type = inv_transform_type
+        # print("computing inverse basis using ", self.inv_transform_type)
+        if self.inv_transform_type is None:
+            ValueError("Inverse transform type not specified")
+        elif self.inv_transform_type == 'pseudo_inverse':
+            self._basis_inv = np.linalg.pinv(self.basis)
+        elif self.inv_transform_type == 'quadrature':
+            if weights is None:
+                print("Warning: No weights specified for quadrature transform,"
+                      "calculating weights using voronoi tessellation of sphere")
+                weights = calculate_sampling_weights(self.coords)
+            self._basis_inv = np.conj(self.basis).T * (weights)
+
+    @property
+    def basis_inv_gradient_theta(self):
+        if self._basis_inv_gradient_theta is None or self._basis_inv_gradient_phi is None:
+            self.compute_inverse_gradient()
+        return self._basis_inv_gradient_theta
+
+    @property
+    def basis_inv_gradient_phi(self):
+        if self._basis_inv_gradient_theta is None or self._basis_inv_gradient_phi is None:
+            self.compute_inverse_gradient()
+        return self._basis_inv_gradient_phi
+
+    def compute_inverse_gradient(self, inv_transform_type=None, weights=None):
+        if inv_transform_type != self.inv_transform_type:
+            self.inv_transform_type = inv_transform_type
+        if self.inv_transform_type is None:
+            ValueError("Inverse transform type not specified")
+        elif self.inv_transform_type == 'pseudo_inverse':
+            self._basis_inv_gradient_theta = np.linalg.pinv(self.basis_gradient_theta)
+            self._basis_inv_gradient_phi = np.linalg.pinv(self.basis_gradient_phi)
+        elif self.inv_transform_type == 'quadrature':
+            if weights is None:
+                weights = calculate_sampling_weights(self.coords)
+            self._basis_inv_gradient_theta = np.conj(self.basis_gradient_theta).T * (4 * np.pi * weights)
+            self._basis_inv_gradient_phi = np.conj(self.basis_gradient_phi).T * (4 * np.pi * weights)
+
+    def transform(self, signal):
```

Comment:
Another general question: Should the transform functions below be part of the SH class or would they go somewhere else? I guess with the inverse defined, they could be part. But aren't there cases where radial functions are combined with the SH matrix before the transform?

Reply:
yes, I would defenetly remove the the transform from the SH class, I think that is a separate method

Reply:
OK!